src/System.Collections/src/System/Collections/Generic/Stack.cs

```diff
@@ -202,69 +202,103 @@ public void TrimExcess()
         // is empty, Peek throws an InvalidOperationException.
         public T Peek()
         {
-            if (_size == 0)
+            int size = _size - 1;
+            T[] array = _array;
```

Comment:
Is this copy really necessary?

Reply:
With the copy `T[] array = _array` it gets:
```asm
000007fe`7bafc5d4 8b4110          mov     eax,dword ptr [rcx+10h]
000007fe`7bafc5d7 ffc8            dec     eax
000007fe`7bafc5d9 488b5108        mov     rdx,qword ptr [rcx+8]
000007fe`7bafc5dd 448b4208        mov     r8d,dword ptr [rdx+8]
000007fe`7bafc5e1 443bc0          cmp     r8d,eax
000007fe`7bafc5e4 760c            jbe     000007fe`7bafc5f2
000007fe`7bafc5e6 4863c0          movsxd  rax,eax
000007fe`7bafc5e9 8b448210        mov     eax,dword ptr [rdx+rax*4+10h]
000007fe`7bafc5ed 4883c428        add     rsp,28h
000007fe`7bafc5f1 c3              ret
```

With out the copy and
```c#
if ((uint)size >= (uint)_array.Length)
...
return _array[size];
```
it gets:
```asm
000007fe`7bb0c5d4 8b4110          mov     eax,dword ptr [rcx+10h]
000007fe`7bb0c5d7 ffc8            dec     eax
000007fe`7bb0c5d9 488b5108        mov     rdx,qword ptr [rcx+8]
000007fe`7bb0c5dd 448b4208        mov     r8d,dword ptr [rdx+8]
000007fe`7bb0c5e1 443bc0          cmp     r8d,eax
000007fe`7bb0c5e4 760c            jbe     000007fe`7bb0c5f2
000007fe`7bb0c5e6 4863c0          movsxd  rax,eax
000007fe`7bb0c5e9 8b448210        mov     eax,dword ptr [rdx+rax*4+10h]
000007fe`7bb0c5ed 4883c428        add     rsp,28h
000007fe`7bb0c5f1 c3              ret
```

So basically the same code. Hence the copy can be avoided.

It seems as the JIT elides bound checks even on field access. Is this new?

Reply:
> It seems as the JIT elides bound checks even on field access. Is this new? (and is this safe?)

It's safe in the sense that the JIT generates correct code. Note that there's only one field load in the generated code. CSE removed the second one so in effect it did the same optimization that you did manually.

The fact that the range check is actually removed might be relatively new, in the past the JIT did not propagate existing analysis information to variables created by CSE and this usually prevented further optimizations. Some improvements in this area were done in https://github.com/dotnet/coreclr/pull/9615

In general I'd recommend avoiding multiple field loads when arrays are involved, CSE may not be able to always eliminate redundant loads and even if it does there may still be cases where it doesn't work as well as manual optimization. 

Still, in trivial cases such as `Peek` it should work fine and thus there's little reason to complicate the code.

Reply:
Thanks for the great reply. :+1:

I'll update the PR for this change.

Reply:
For reference in `TryPeek` this JIT-optimization doesn't apply:

### With Copy
```asm
000007fe`7b86c610 8b4110          mov     eax,dword ptr [rcx+10h]
000007fe`7b86c613 448d40ff        lea     r8d,[rax-1]
000007fe`7b86c617 488b4908        mov     rcx,qword ptr [rcx+8]
000007fe`7b86c61b 44394108        cmp     dword ptr [rcx+8],r8d
000007fe`7b86c61f 7705            ja      000007fe`7b86c626
000007fe`7b86c621 33c0            xor     eax,eax
000007fe`7b86c623 8902            mov     dword ptr [rdx],eax
000007fe`7b86c625 c3              ret
000007fe`7b86c626 4963c0          movsxd  rax,r8d
000007fe`7b86c629 8b448110        mov     eax,dword ptr [rcx+rax*4+10h]
000007fe`7b86c62d 8902            mov     dword ptr [rdx],eax
000007fe`7b86c62f b801000000      mov     eax,1
```

### Without Copy
```asm
000007fe`7b87c610 8b4110          mov     eax,dword ptr [rcx+10h]
000007fe`7b87c613 448d40ff        lea     r8d,[rax-1]
000007fe`7b87c617 488b4108        mov     rax,qword ptr [rcx+8]
000007fe`7b87c61b 8b4008          mov     eax,dword ptr [rax+8]
000007fe`7b87c61e 413bc0          cmp     eax,r8d
000007fe`7b87c621 7705            ja      000007fe`7b87c628
000007fe`7b87c623 33c0            xor     eax,eax
000007fe`7b87c625 8902            mov     dword ptr [rdx],eax
000007fe`7b87c627 c3              ret
000007fe`7b87c628 488b4108        mov     rax,qword ptr [rcx+8]
000007fe`7b87c62c 4963c8          movsxd  rcx,r8d
000007fe`7b87c62f 8b448810        mov     eax,dword ptr [rax+rcx*4+10h]
000007fe`7b87c633 8902            mov     dword ptr [rdx],eax
000007fe`7b87c635 b801000000      mov     eax,1
```

Reply:
Hmm, interesting. The range check seems to have been eliminated but there are loads of the `_array` field. That doesn't seem right, I'll have to look into this.

Reply:
I'm not yet sure what's going on but the generated code is definitely incorrect. Created https://github.com/dotnet/coreclr/issues/15671